# COMP5318 Assignment 1: Rice Classification

##### Group number: 92
##### Student 1 SID: 550455842
##### Student 2 SID: 550306979
##### Student 3 SID: 550293521

## **1. Data Pre-processing**

### 1.1 Libraries

In [128]:
import pandas as pd
from sklearn.model_selection import StratifiedKFold
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import MinMaxScaler
from warnings import simplefilter

### 1.2 Ignore future warnings

In [129]:
simplefilter(action='ignore', category=FutureWarning)

### 1.3 Function

In [137]:
def read_csv(filename):
    return pd.read_csv(filename, na_values='?')

def check_values(dataset):
    return dataset.head()

def check_null_values(dataset):
    return dataset.isnull().sum()

def fill_missing_values(dataset):
    imputer = SimpleImputer(strategy='mean')

    feature_cols = dataset.columns.drop('class')

    dataset_inputed = dataset.copy()

    dataset_inputed[feature_cols] = imputer.fit_transform(dataset[feature_cols])

    return dataset_inputed

def check_min_max(dataset):
    feature_cols = dataset.columns.drop('class')

    return dataset[feature_cols].min(), dataset[feature_cols].max()

def normalise(dataset):
    scaler = MinMaxScaler()

    feature_cols = dataset.columns.drop('class')

    dataset_normalised = dataset.copy()

    dataset_normalised[feature_cols] = scaler.fit_transform(dataset[feature_cols])

    return dataset_normalised

def check_unique_values(dataset):
    return dataset['class'].unique()

def mapping(dataset):
    dataset_mapped = dataset.copy()

    dataset_mapped['class'] = dataset['class'].map({
        'class1': 0,
        'class2': 1
    })

    return dataset_mapped

# Print first ten rows of pre-processed dataset to 4 decimal places as per assignment spec
# A function is provided to assist
def print_data(X, y, n_rows=10):
    """Takes a numpy data array and target and prints the first ten rows.
    
    Arguments:
        X: numpy array of shape (n_examples, n_features)
        y: numpy array of shape (n_examples)
        n_rows: numpy of rows to print
    """
    for example_num in range(n_rows):
        for feature in X[example_num]:
            print("{:.4f}".format(feature), end=",")

        if example_num == len(X)-1:
            print(y[example_num],end="")
        else:
            print(y[example_num])

def convert2numpy(dataset):
    X = dataset.drop(columns=['class']).to_numpy()
    y = dataset['class'].to_numpy()
    return X, y

### 1.4 Read dataset

In [138]:
rice_dataset = read_csv('rice-final2.csv')

### 1.5 Filling in the missing attribute values 

#### 1.5.1 Check values

In [139]:
check_values(rice_dataset)

,Area,Perimiter,Major_Axis_Length,Minor_Axis_Length,Eccentricity,Convex_Area,Extent,class
0,12573.0,461.466003,192.903351,84.572075,0.898772,12893.0,0.550433,class2
1,12845.0,464.121002,194.332214,85.524338,0.897952,13125.0,0.774962,class2
2,14055.0,488.748993,207.751755,87.250328,0.907536,14484.0,0.550076,class1
3,14412.0,490.324005,207.476135,89.689514,0.901735,14703.0,0.598853,class1
4,14658.0,477.117004,189.566635,99.997780,0.849551,15048.0,0.649504,class2


#### 1.5.2 Check null values of each columns

In [143]:
check_null_values(rice_dataset)

Area                 4
Perimiter            4
Major_Axis_Length    5
Minor_Axis_Length    3
Eccentricity         6
Convex_Area          5
Extent               2
class                0
dtype: int64

#### 1.5.3 Filling in the missing attribute values

In [141]:
rice_imputed = fill_missing_values(rice_dataset)

In [142]:
check_null_values(rice_imputed)

Area                 0
Perimiter            0
Major_Axis_Length    0
Minor_Axis_Length    0
Eccentricity         0
Convex_Area          0
Extent               0
class                0
dtype: int64

### 1.6 Normalising	the	data

#### 1.6.1 Check min and max values

In [155]:
rice_min, rice_max = check_min_max(rice_imputed)

In [156]:
rice_min

Area                 7943.000000
Perimiter             359.100006
Major_Axis_Length     145.264465
Minor_Axis_Length      63.344753
Eccentricity            0.799511
Convex_Area          8080.000000
Extent                  0.511160
dtype: float64

In [157]:
rice_max

Area                 17948.000000
Perimiter              548.445984
Major_Axis_Length      238.435089
Minor_Axis_Length      107.542450
Eccentricity             0.934006
Convex_Area          18322.000000
Extent                   0.839661
dtype: float64

#### 1.6.2 normalising

In [159]:
rice_norm = normalise(rice_imputed)

#### 1.6.3 Check min and max values again

In [160]:
rice_norm_min, rice_norm_max = check_min_max(rice_norm)

In [161]:
rice_norm_min

Area                 0.0
Perimiter            0.0
Major_Axis_Length    0.0
Minor_Axis_Length    0.0
Eccentricity         0.0
Convex_Area          0.0
Extent               0.0
dtype: float64

In [162]:
rice_norm_max

Area                 1.0
Perimiter            1.0
Major_Axis_Length    1.0
Minor_Axis_Length    1.0
Eccentricity         1.0
Convex_Area          1.0
Extent               1.0
dtype: float64

### 1.7 Changing the class values

#### 1.7.1 Check values

In [167]:
check_unique_values(rice_norm)

array(['class2', 'class1'], dtype=object)

#### 1.7.2 Mapping

In [165]:
rice_mapped = mapping(rice_norm)

#### 1.7.3 Check values again

In [166]:
check_unique_values(rice_mapped)

array([1, 0])

### 1.8 Print the first	10 rows of the pre-processed dataset

In [169]:
X, y = convert2numpy(rice_mapped)

In [171]:
print_data(X, y)

0.4628,0.5406,0.5113,0.4803,0.7380,0.4699,0.1196,1
0.4900,0.5547,0.5266,0.5018,0.7319,0.4926,0.8030,1
0.6109,0.6847,0.6707,0.5409,0.8032,0.6253,0.1185,0
0.6466,0.6930,0.6677,0.5961,0.7601,0.6467,0.2669,0
0.6712,0.6233,0.4755,0.8293,0.3721,0.6803,0.4211,1
0.2634,0.2932,0.2414,0.4127,0.5521,0.2752,0.2825,1
0.8175,0.9501,0.9515,0.5925,0.9245,0.8162,0.0000,0
0.3174,0.3588,0.3601,0.3908,0.6921,0.3261,0.8510,1
0.3130,0.3050,0.2150,0.5189,0.3974,0.3159,0.4570,1
0.5120,0.5237,0.4409,0.6235,0.5460,0.5111,0.3155,1


## **2. Build Classifiers**

- Part 1:  Logistic Regression, Naïve Bayes
- Part 2:  KNN, Decision Tree, Ada Boost, Gradient Boost, Random Forest, SVM

### 2.1 Preparation

#### 2.1.1 Libraries

In [179]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import accuracy_score, f1_score
from sklearn.naive_bayes import GaussianNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import (
    AdaBoostClassifier,
    GradientBoostingClassifier,
    RandomForestClassifier
)
from sklearn.svm import SVC

#### 2.1.2 Function

In [180]:
def cv_kfold(n_splits=10, shuffle=True, random_state=0):
    return StratifiedKFold(
        n_splits=n_splits, 
        shuffle=shuffle, 
        random_state=random_state
    )

def logistic_regression(X, y, cvkfold, scoring = 'accuracy'):
    lr = LogisticRegression(
        max_iter=1000,
        random_state=0
    )

    lr_scores = cross_val_score(
        lr,
        X,
        y,
        cv=cvkfold,
        scoring=scoring
    )

    return lr_scores.mean()

def naive_bayes(X, y, cvkfold, scoring = 'accuracy'):
    nb = GaussianNB()

    nb_scores = cross_val_score(
        nb,
        X,
        y,
        cv=cvkfold,
        scoring=scoring
    )

    return nb_scores.mean()

### **2.2 Part 1: Cross-validation without parameter tuning**

#### 2.2.1 CV K-fold

In [ ]:
cvKFold = cv_kfold()

#### 2.2.2 Logistic Regression

In [175]:
lr_cv_accuracy = logistic_regression(X, y, cvKFold)

#### 2.2.3 Naïve Bayes

In [177]:
nb_cv_accuracy = naive_bayes(X, y, cvKFold)

#### 2.2.4 Part 1 Results


In [178]:
# Print results for each classifier in part 1 to 4 decimal places here:
print(f"LogR average cross-validation accuracy: {lr_cv_accuracy:.4f}")
print(f"NB average cross-validation accuracy: {nb_cv_accuracy:.4f}")

LogR average cross-validation accuracy: 0.9386
NB average cross-validation accuracy: 0.9264


### **2.3 Part 2: Cross-validation with parameter tuning**

#### 2.3.1 Function

In [183]:

def my_train_test_split(X, y, test_size=0.2, random_state=0):
    X_train, X_test, y_train, y_test = train_test_split(
        X,
        y,
        test_size=test_size,
        stratify=y,
        random_state=random_state
    )

    return X_train, X_test, y_train, y_test

def knn(k, p, cvkfold, X_train, y_train, X_test, y_test, scoring = 'accuracy'):
    knn = KNeighborsClassifier()

    knn_param_grid = {
        'n_neighbors': k,
        'p': p
    }

    knn_grid = GridSearchCV(
        estimator=knn,
        param_grid=knn_param_grid,
        scoring=scoring,
        cv=cvkfold
    )

    knn_grid.fit(X_train, y_train)

    knn_pred = knn_grid.predict(X_test)

    knn_test_accuracy = accuracy_score(y_test, knn_pred)

    return knn_grid, knn_test_accuracy

def decision_tree(X_train, y_train, X_test, y_test, max_depth, min_samples_split, min_samples_leaf, cvkfold, random_state = 0, scoring = 'accuracy'):
    dt = DecisionTreeClassifier(random_state=random_state)

    dt_param_grid = {
        'max_depth': max_depth,
        'min_samples_split': min_samples_split,
        'min_samples_leaf': min_samples_leaf
    }

    dt_grid = GridSearchCV(
        estimator=dt,
        param_grid=dt_param_grid,
        scoring=scoring,
        cv=cvkfold
    )

    dt_grid.fit(X_train, y_train)

    dt_pred = dt_grid.predict(X_test)

    dt_test_accuracy = accuracy_score(y_test, dt_pred)

    return dt_grid, dt_test_accuracy

def ada_boost(X_train, y_train, X_test, y_test, n_estimators, learning_rate, cvkfold, scoring = 'accuracy', random_state = 0):
    ada_param_grid = {
        'n_estimators': n_estimators,
        'learning_rate': learning_rate
    }

    ada_grid = GridSearchCV(
        AdaBoostClassifier(random_state=random_state),
        ada_param_grid,
        scoring=scoring,
        cv=cvkfold
    )

    ada_grid.fit(X_train, y_train)
    ada_pred = ada_grid.predict(X_test)
    ada_test_accuracy = accuracy_score(y_test, ada_pred)

    return ada_grid, ada_test_accuracy

def gradient_boost(X_train, y_train, X_test, y_test, max_depth, n_estimators, learning_rate, cvkfold, scoring = 'accuracy', random_state = 0):
    gb_param_grid = {
        'n_estimators': n_estimators,
        'learning_rate': learning_rate,
        'max_depth': max_depth
    }

    gb_grid = GridSearchCV(
        GradientBoostingClassifier(random_state=random_state),
        gb_param_grid,
        scoring=scoring,
        cv=cvkfold
    )

    gb_grid.fit(X_train, y_train)
    gb_pred = gb_grid.predict(X_test)
    gb_test_accuracy = accuracy_score(y_test, gb_pred)

    return gb_grid, gb_test_accuracy

def random_forest(X_train, y_train, X_test, y_test, n_estimators, max_leaf_nodes, cvkfold, scoring = 'accuracy', random_state = 0):
    rf_param_grid = {
        'n_estimators': n_estimators,
        'max_leaf_nodes': max_leaf_nodes
    }

    rf_grid = GridSearchCV(
        RandomForestClassifier(random_state=random_state),
        rf_param_grid,
        scoring=scoring,
        cv=cvkfold
    )

    rf_grid.fit(X_train, y_train)
    rf_pred = rf_grid.predict(X_test)

    rf_test_accuracy = accuracy_score(y_test, rf_pred)
    rf_macro_f1 = f1_score(y_test, rf_pred, average='macro')
    rf_weighted_f1 = f1_score(y_test, rf_pred, average='weighted')

    return rf_grid, rf_test_accuracy, rf_macro_f1, rf_weighted_f1

def svm(X_train, y_train, X_test, y_test, C, gamma, kernel, cvkfold, scoring = 'accuracy'):
    svm_param_grid = {
        'C': C,
        'kernel': kernel,
        'gamma': gamma
    }

    svm_grid = GridSearchCV(
        SVC(),
        svm_param_grid,
        scoring=scoring,
        cv=cvkfold
    )

    svm_grid.fit(X_train, y_train)
    svm_pred = svm_grid.predict(X_test)
    svm_test_accuracy = accuracy_score(y_test, svm_pred)

    return svm_grid, svm_test_accuracy

#### 2.3.2 Train/Test split

In [182]:
X_train, X_test, y_train, y_test = my_train_test_split(X, y)

#### 2.3.3 KNN

In [184]:
# KNN 
# parameters you may consider
k = [1, 3, 5, 7]
p = [1, 2]

knn_grid, knn_test_accuracy = knn(k, p, cvKFold, X_train, y_train, X_test, y_test)

#### 2.3.4 Decision Tree 

In [185]:
# Decision Tree 
# parameters you may consider
max_depth = [3, 5, 7, 10]
min_samples_split = [2, 5, 10]
min_samples_leaf = [1, 2, 4]

dt_grid, dt_test_accuracy = decision_tree(X_train, y_train, X_test, y_test, max_depth, min_samples_split, min_samples_leaf, cvKFold)

### 2.3.5 Ada Boost

In [186]:
# Ada Boost
# parameters you may consider
n_estimators = [50, 100, 150]
learning_rate = [0.1, 0.2, 0.3, 0.5]

ada_grid, ada_test_accuracy = ada_boost(X_train, y_train, X_test, y_test, n_estimators, learning_rate, cvKFold)

#### 2.3.6 Gradient Boost

In [187]:
# Gradient Boost
# parameters you may consider
max_depth = [1, 3, 5, 7]
n_estimators = [50, 100, 150]
learning_rate = [0.1, 0.2, 0.3, 0.5]

gb_grid, gb_test_accuracy = gradient_boost(X_train, y_train, X_test, y_test, max_depth, n_estimators, learning_rate, cvKFold)


#### 2.3.7 Random Forest

In [188]:
# Random Forest
# You should use RandomForestClassifier from sklearn.ensemble with information gain and max_features set to ‘sqrt’.
# parameters you may consider
n_estimators = [10, 30, 60, 100]
max_leaf_nodes = [6, 12]

rf_grid, rf_test_accuracy, rf_macro_f1, rf_weighted_f1 = random_forest(X_train, y_train, X_test, y_test, n_estimators, max_leaf_nodes, cvKFold)

#### 2.3.8 SVM

In [189]:
# SVM
# parameters you may consider
C = [0.01, 0.1, 1, 5]
gamma = [0.01, 0.1, 1, 10]
# optional
kernel = ['linear', 'rbf']

svm_grid, svm_test_accuracy = svm(X_train, y_train, X_test, y_test, C, gamma, kernel, cvKFold)


#### 2.3.9 Part 2: Results

In [190]:
# Perform Grid Search with 10-fold stratified cross-validation (GridSearchCV in sklearn). 
# The stratified folds from cvKFold should be provided to GridSearchV

# This should include using train_test_split from sklearn.model_selection with stratification and random_state=0
# Print results for each classifier here. All the reported results should be printed to 4 decimal places except for the integers such as "k", "p", n_estimators" and "max_leaf_nodes".

# example printing:
print("KNN best k: ", knn_grid.best_params_['n_neighbors'])
print("KNN best p: ", knn_grid.best_params_['p'])
print(f"KNN cross-validation accuracy: {knn_grid.best_score_:.4f}")
print(f"KNN test set accuracy: {knn_test_accuracy:.4f}")
print()

print("Decision Tree best parameters:", dt_grid.best_params_)
print(f"Decision Tree cross-validation accuracy: {dt_grid.best_score_:.4f}")
print(f"Decision Tree test set accuracy: {dt_test_accuracy:.4f}")
print()

print("AdaBoost best parameters:", ada_grid.best_params_)
print(f"AdaBoost cross-validation accuracy: {ada_grid.best_score_:.4f}")
print(f"AdaBoost test set accuracy: {ada_test_accuracy:.4f}")
print()

print("Gradient Boosting best parameters:", gb_grid.best_params_)
print(f"Gradient Boosting cross-validation accuracy: {gb_grid.best_score_:.4f}")
print(f"Gradient Boosting test set accuracy: {gb_test_accuracy:.4f}")
print()

print("RF best n_estimators: ", rf_grid.best_params_['n_estimators'])
print("RF best max_leaf_nodes: ", rf_grid.best_params_['max_leaf_nodes'])
print(f"RF cross-validation accuracy: {rf_grid.best_score_:.4f}")
print(f"RF test set accuracy: {rf_test_accuracy:.4f}")
print(f"RF test set macro average F1: {rf_macro_f1:.4f}")
print(f"RF test set weighted average F1: {rf_weighted_f1:.4f}")
print()

print("SVM best parameters:", svm_grid.best_params_)
print(f"SVM cross-validation accuracy: {svm_grid.best_score_:.4f}")
print(f"SVM test set accuracy: {svm_test_accuracy:.4f}")

KNN best k:  7
KNN best p:  2
KNN cross-validation accuracy: 0.9375
KNN test set accuracy: 0.9250

Decision Tree best parameters: {'max_depth': 3, 'min_samples_leaf': 1, 'min_samples_split': 2}
Decision Tree cross-validation accuracy: 0.9357
Decision Tree test set accuracy: 0.9429

AdaBoost best parameters: {'learning_rate': 0.2, 'n_estimators': 150}
AdaBoost cross-validation accuracy: 0.9455
AdaBoost test set accuracy: 0.9429

Gradient Boosting best parameters: {'learning_rate': 0.1, 'max_depth': 1, 'n_estimators': 50}
Gradient Boosting cross-validation accuracy: 0.9446
Gradient Boosting test set accuracy: 0.9429

RF best n_estimators:  10
RF best max_leaf_nodes:  6
RF cross-validation accuracy: 0.9402
RF test set accuracy: 0.9429
RF test set macro average F1: 0.9415
RF test set weighted average F1: 0.9428

SVM best parameters: {'C': 5, 'gamma': 1, 'kernel': 'rbf'}
SVM cross-validation accuracy: 0.9429
SVM test set accuracy: 0.9321


### Test your code

In [125]:
#load the test dataset to test out your model 

## **3. Reflection and Discussion**



All six classifiers performed well, with test accuracies between 0.9250 and 0.9429. AdaBoost had the highest cross-validation accuracy at 0.9455, followed by Gradient Boosting and SVM. On the test set, Decision Tree, AdaBoost, Gradient Boosting and Random Forest all achieved the highest accuracy of 0.9429. KNN had the lowest test accuracy at 0.9250.

The cross-validation and test results were generally close, which suggests that most models generalised well. AdaBoost and Gradient Boosting were especially stable, with only small differences between their cross-validation and test accuracies. Random Forest also performed consistently. Its macro F1-score of 0.9415 and weighted F1-score of 0.9428 were close to its test accuracy, suggesting balanced performance across classes.

Hyperparameter tuning helped select suitable model settings and control model complexity. For example, the best Decision Tree used a maximum depth of 3, showing that a relatively simple tree was enough. Gradient Boosting used shallow trees with a maximum depth of 1, while AdaBoost used 150 estimators with a learning rate of 0.2. These settings were selected using cross-validation to improve generalisation.

For KNN, the best setting was `k=7` and `p=2`, but it still had the lowest test accuracy. SVM selected an RBF kernel with `C=5` and `gamma=1`, although its test accuracy was slightly lower than its cross-validation accuracy.

Overall, the ensemble methods performed very well. AdaBoost achieved the highest cross-validation accuracy and also matched the highest test accuracy, so it can be considered the strongest overall model. Gradient Boosting and Random Forest were also strong alternatives.

One limitation is that these results mainly show performance after tuning. Without comparing them with the default model settings, we cannot measure exactly how much hyperparameter tuning improved each classifier.


## **AI Acknowledgement**

Generative AI (ChatGPT, OpenAI) was used to assist with understanding the assignment requirements, explaining machine learning concepts and scikit-learn functions, reviewing code structure, and improving the clarity and wording of the reflection and discussion. All code was implemented and run by our group, and the final results, analysis and submitted content were reviewed and verified by us.